File: 230. Kth Smallest Element in a BST/230. Kth Smallest Element in a BST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  int res;
  int count;

  public int kthSmallest(TreeNode root, int k) {
    count = k;
    dfs(root);
    return res;
  }

  private void dfs(TreeNode node) {
    if (node == null || count == 0) {
      return;
    }

    dfs(node.left);

    count--;
    if (count == 0) {
      res = node.val;
      return;
    }

    dfs(node.right);
  }
}

File: 230. Kth Smallest Element in a BST/230. Kth Smallest Element in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function (root, k) {
  res = null;
  count = k;

  const dfs = (node) => {
    if (!node || res) {
      return;
    }

    dfs(node.left);

    count--;
    if (count == 0) {
      res = node.val;
      return;
    }

    dfs(node.right);
  };

  dfs(root);
  return res;
};
